### ISSUE TITLE
No enhance with decoding_enhanced.py

### ISSUE THREAD

**Issue by nicolas-ivanov** [LeavesBreathe/Seq2Seq_Upgrade_TensorFlow#5]
Dear Leaves,
thank you for trying to bring diversification to the sequence decoding procedure, however I believe it doesn't work that simple with seq2seq. The approach you took is aweseome while predicting a sequence word by word or character by character as indeed you can see in [Karpathy's code](https://github.com/karpathy/char-rnn/blob/4297a9bf69726823d944ad971555e91204f12ca8/sample.lua) or in [keras example](https://github.com/fchollet/keras/blob/master/examples/lstm_text_generation.py):

```python
def sample(a, temperature=1.0):
    # helper function to sample an index from a probability array
    a = np.log(a) / temperature
    a = np.exp(a) / np.sum(np.exp(a))
    return np.argmax(np.random.multinomial(1, a, 1))
```
However trying out such an approach with [our code](https://github.com/nicolas-ivanov/tf_seq2seq_chatbot) only worsen results: the generated sequence just becomes grammatically inconsistent when trying to pick a word that is not the most probable one.
Do you have other experience?

Another argument against such diversification in seq2seq is the following article: http://arxiv.org/abs/1510.03055
Guys from Stanford and Microsoft had to invent quite a lot to improve their seq2seq diversity...

**Comment by LeavesBreathe** [comment]
@nicolas-ivanov , you are completely correct. This is just a basic sampling over the *already* produced sequence. I tried it out a little bit, and it helped me some (when I set temperature to 0.8). 

In the future, I'm going to work on temperature sampling *before* the next timestep is generated. I'll add that to the list. Its something I definitely want to do. 

I didn't mean to mislead you or anything, I just wanted to get some basic temperature mod out.
